src/resnet.jl

```diff
@@ -1,97 +1,271 @@
+# Arguments
+- `block`: a function with input `(inplanes, outplanes downsample=false)` that returns
+           a new residual block (see [`Metalhead.basicblock`](#) and [`Metalhead.bottleneck`](#))
+- `shortcut_config`: the type of shortcut style (either `:A`, `:B`, or `:C`)
+- `channel_config`: the growth rate of the output feature maps within a residual block
+- `block_config`: a list of the number of residual blocks at each stage
+- `nclasses`: the number of output classes
+"""
+function resnet(; block, shortcut_config, channel_config, block_config, nclasses=1000)
+  inplanes = 64
+  baseplanes = 64
+  layers = []
+  append!(layers, conv_bn((7, 7), 3, inplanes; stride=2, pad=(3, 3)))
+  push!(layers, MaxPool((3, 3), stride=(2, 2), pad=(1, 1)))
+  for (i, nrepeats) in enumerate(block_config)
+    outplanes = baseplanes .* channel_config
+    if shortcut_config == :A
+      push!(layers, Parallel(+, block(inplanes, outplanes, i != 1),
+                                skip_identity(inplanes, outplanes)))
+    elseif shortcut_config == :B || shortcut_config == :C
+      push!(layers, Parallel(+, block(inplanes, outplanes, i != 1),
+                                skip_projection(inplanes, outplanes[end], i != 1)))
+    end
+    inplanes = outplanes[end]
+    for j in 2:nrepeats
+      if shortcut_config == :A || shortcut_config == :B
+        push!(layers, Parallel(+, block(inplanes, outplanes, false),
+                                  skip_identity(inplanes, outplanes[end])))
+      elseif shortcut_config == :C
+        push!(layers, Parallel(+, block(inplanes, outplanes, false),
+                                  skip_projection(inplanes, outplanes, false)))
+      end
+      inplanes = outplanes[end]
     end
-    initial_filters *= 2
+    baseplanes *= 2
   end
+  push!(layers, AdaptiveMeanPool((1, 1)))
+  push!(layers, flatten)
+  push!(layers, Dense(inplanes, nclasses))
+
+  return Chain(layers...)
+end
+
+const resnet_config =
+  Dict("resnet18" => ([1, 1], [2, 2, 2, 2]),
+       "resnet34" => ([1, 1], [3, 4, 6, 3]),
+       "resnet50" => ([1, 1, 4], [3, 4, 6, 3]),
+       "resnet101" => ([1, 1, 4], [3, 4, 23, 3]),
+       "resnet152" => ([1, 1, 4], [3, 8, 36, 3]))
+
+"""
+    ResNet{BC, SC}(channel_config, block_config; nclasses=1000)
 
-  push!(layer_arr, MeanPool((7,7)))
-  push!(layer_arr, x -> reshape(x, :, size(x,4)))
-  push!(layer_arr, (Dense(2048, 1000)))
-  push!(layer_arr, softmax)
+Create a `ResNet` model
+([reference](https://arxiv.org/abs/1512.03385v1)).
+See also [`resnet`](#).
 
-  Chain(layer_arr...)
+# Arguments
+- `BC`: residual block style (either `:basic` or `:bottleneck`)
+- `SC`: shortcut style (either `:A` or `:B`)
+- `channel_config`: the growth rate of the output feature maps within a residual block
+- `block_config`: a list of the number of residual blocks at each stage
+- `nclasses`: the number of output classes
+"""
+struct ResNet{BC, SC, T}
+  layers::T
 end
```

Comment:
With this struct outermost, I think you will want to overload `show`, behind `if isdefined(Flux, :_big_show)`. It's a nice tests case:

<img width="1048" alt="Screenshot 2021-07-10 at 15 18 24" src="https://user-images.githubusercontent.com/32575566/125174293-3a07da00-e192-11eb-99d4-22011a5fdc5c.png">